src/Icicle/Source/Checker/Constraint.hs

```diff
@@ -171,10 +171,13 @@ generateQ qq@(Query (c:_) _)
     -- >    : Aggregate a
     GroupFold _ k v x
      -> do  (x',sx)    <- generateX x
+
             retKey     <- Temporality TemporalityElement . TypeVar <$> fresh
             retVal     <- Temporality TemporalityElement . TypeVar <$> fresh
+            removeElementBinds
             (q',sq,t') <- withBind k retKey
                         $ withBind v retVal rest
+
```

Comment:
yeah, I was a bit unsure about what to do after this. Do I re-add the things that were removed...?

I can try and kill the state monad, just pass both the environment and constraint sets around... Dubious honour :imp: